File: src/android/ansight-core/src/main/kotlin/ai/ansight/runtime/TouchVisualTreeCaptureCoordinator.kt
package ai.ansight.runtime

import java.util.UUID
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit

internal enum class TouchVisualTreeGesturePhase(val wireName: String) {
    Started("started"),
    Checkpoint("checkpoint"),
    Ended("ended"),
}

internal data class TouchVisualTreeCaptureTrigger(
    val gestureId: String,
    val touchAction: String,
    val gesturePhase: TouchVisualTreeGesturePhase,
    val touchCapturedAtUtc: String,
)

internal class TouchVisualTreeCaptureCoordinator(
    private val capture: (TouchVisualTreeCaptureTrigger) -> Unit,
    private val minimumCaptureIntervalMilliseconds: Long = DefaultMinimumCaptureIntervalMilliseconds,
) : AutoCloseable {
    init {
        require(minimumCaptureIntervalMilliseconds > 0)
    }

    private val lock = Any()
    private val activePointerIds = mutableSetOf<Long>()
    private val executor: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor { runnable ->
        Thread(runnable, "AnsightTouchVisualTree").apply { isDaemon = true }
    }
    private var pendingTrigger: TouchVisualTreeCaptureTrigger? = null
    private var workerScheduled = false
    private var nextCaptureAllowedAtNanoseconds = 0L
    private var gestureId: String? = null
    private var closed = false

    fun observe(touch: CapturedTouch) {
        var trigger: TouchVisualTreeCaptureTrigger? = null
        synchronized(lock) {
            if (closed) {
                return
            }

            when (touch.action.lowercase()) {
                "down" -> {
                    val beginsGesture = activePointerIds.isEmpty()
                    activePointerIds += touch.pointerId
                    if (beginsGesture) {
                        gestureId = "gesture-${UUID.randomUUID()}"
                    }
                    trigger = createTrigger(
                        touch,
                        if (beginsGesture) {
                            TouchVisualTreeGesturePhase.Started
                        } else {
                            TouchVisualTreeGesturePhase.Checkpoint
                        },
                    )
                }
                "move" -> activePointerIds += touch.pointerId
                "up" -> {
                    activePointerIds -= touch.pointerId
                    val phase = if (activePointerIds.isEmpty()) {
                        TouchVisualTreeGesturePhase.Ended
                    } else {
                        TouchVisualTreeGesturePhase.Checkpoint
                    }
                    trigger = createTrigger(touch, phase)
                }
                "cancel" -> {
                    activePointerIds.clear()
                    gestureId = null
                }
            }
        }

        trigger?.let(::enqueue)
    }

    override fun close() {
        synchronized(lock) {
            if (closed) {
                return
            }
            closed = true
            activePointerIds.clear()
            pendingTrigger = null
            gestureId = null
        }
        executor.shutdownNow()
    }

    fun interruptGesture() {
        synchronized(lock) {
            activePointerIds.clear()
            pendingTrigger = null
            gestureId = null
        }
    }

    private fun enqueue(trigger: TouchVisualTreeCaptureTrigger) {
        var scheduleDelayMilliseconds = -1L
        synchronized(lock) {
            if (closed) {
                return
            }

            pendingTrigger = selectPendingTrigger(pendingTrigger, trigger)
            if (!workerScheduled) {
                workerScheduled = true
                scheduleDelayMilliseconds = remainingCaptureDelayMilliseconds()
            }
        }

        if (scheduleDelayMilliseconds >= 0) {
            scheduleWorker(scheduleDelayMilliseconds)
        }
    }

    private fun scheduleWorker(delayMilliseconds: Long) {
        runCatching {
            executor.schedule(
                { drainPendingTrigger() },
                delayMilliseconds,
                TimeUnit.MILLISECONDS,
            )
        }.onFailure {
            synchronized(lock) {
                workerScheduled = false
            }
        }
    }

    private fun drainPendingTrigger() {
        val trigger = synchronized(lock) {
            if (closed) {
                workerScheduled = false
                return
            }

            pendingTrigger.also { pendingTrigger = null }
        }

        if (trigger == null) {
            synchronized(lock) {
                workerScheduled = false
            }
            return
        }

        runCatching { capture(trigger) }
        synchronized(lock) {
            nextCaptureAllowedAtNanoseconds = System.nanoTime() +
                TimeUnit.MILLISECONDS.toNanos(minimumCaptureIntervalMilliseconds)
        }

        val shouldContinue = synchronized(lock) {
            if (closed || pendingTrigger == null) {
                workerScheduled = false
                false
            } else {
                true
            }
        }
        if (shouldContinue) {
            scheduleWorker(remainingCaptureDelayMilliseconds())
        }
    }

    private fun remainingCaptureDelayMilliseconds(): Long {
        val remainingNanoseconds = synchronized(lock) {
            nextCaptureAllowedAtNanoseconds - System.nanoTime()
        }
        return if (remainingNanoseconds <= 0) {
            0
        } else {
            TimeUnit.NANOSECONDS.toMillis(remainingNanoseconds - 1) + 1
        }
    }

    private fun selectPendingTrigger(
        pending: TouchVisualTreeCaptureTrigger?,
        incoming: TouchVisualTreeCaptureTrigger,
    ): TouchVisualTreeCaptureTrigger {
        if (pending == null || incoming.gesturePhase == TouchVisualTreeGesturePhase.Started) {
            return incoming
        }
        if (pending.gesturePhase == TouchVisualTreeGesturePhase.Started) {
            return pending
        }
        return if (incoming.gesturePhase == TouchVisualTreeGesturePhase.Ended) incoming else pending
    }

    private fun createTrigger(
        touch: CapturedTouch,
        phase: TouchVisualTreeGesturePhase,
    ): TouchVisualTreeCaptureTrigger = TouchVisualTreeCaptureTrigger(
        gestureId = gestureId ?: "gesture-${UUID.randomUUID()}",
        touchAction = touch.action.lowercase(),
        gesturePhase = phase,
        touchCapturedAtUtc = touch.capturedAtUtc,
    )

    companion object {
        const val DefaultMinimumCaptureIntervalMilliseconds = 750L
    }
}

File: src/dotnet/Ansight.Core/Screenshot/TouchVisualTreeCaptureCoordinator.cs
using System.Diagnostics;
using Ansight.Input;

namespace Ansight.Screenshot;

internal enum TouchVisualTreeGesturePhase
{
    Started,
    Checkpoint,
    Ended
}

internal sealed record TouchVisualTreeCaptureTrigger(
    string GestureId,
    CapturedTouchAction TouchAction,
    TouchVisualTreeGesturePhase GesturePhase,
    DateTimeOffset TouchCapturedAtUtc);

internal sealed class TouchVisualTreeCaptureCoordinator : IDisposable
{
    internal static readonly TimeSpan DefaultMinimumCaptureInterval = TimeSpan.FromMilliseconds(750);

    private readonly Func<TouchVisualTreeCaptureTrigger, CancellationToken, Task> captureAsync;
    private readonly TimeSpan minimumCaptureInterval;
    private readonly Lock stateLock = new();
    private readonly SemaphoreSlim signal = new(0, 1);
    private readonly HashSet<long> activePointerIds = [];
    private TouchCaptureHub? touchCaptureHub;
    private EventHandler<TouchCapturedEventArgs>? touchCapturedHandler;
    private EventHandler? runtimeCaptureInterruptedHandler;
    private EventHandler<AppLifecycleStateChangedEventArgs>? appLifecycleStateChangedHandler;
    private CancellationTokenSource? captureCts;
    private Task? captureTask;
    private TouchVisualTreeCaptureTrigger? pendingTrigger;
    private string? gestureId;
    private bool disposed;

    public TouchVisualTreeCaptureCoordinator(
        Func<TouchVisualTreeCaptureTrigger, CancellationToken, Task> captureAsync,
        TimeSpan? minimumCaptureInterval = null)
    {
        this.captureAsync = captureAsync ?? throw new ArgumentNullException(nameof(captureAsync));
        this.minimumCaptureInterval = minimumCaptureInterval.GetValueOrDefault(DefaultMinimumCaptureInterval);
        if (this.minimumCaptureInterval <= TimeSpan.Zero)
        {
            throw new ArgumentOutOfRangeException(nameof(minimumCaptureInterval));
        }
    }

    public void Start(TouchCaptureHub touchCaptureHub)
    {
        ArgumentNullException.ThrowIfNull(touchCaptureHub);
        ObjectDisposedException.ThrowIf(disposed, this);

        lock (stateLock)
        {
            if (captureCts is not null)
            {
                return;
            }

            this.touchCaptureHub = touchCaptureHub;
            captureCts = new CancellationTokenSource();
            touchCapturedHandler = (_, args) => Observe(args.Touch);
            runtimeCaptureInterruptedHandler = (_, _) => ResetGesture();
            appLifecycleStateChangedHandler = (_, args) =>
            {
                if (args.State == AppLifecycleState.Background)
                {
                    ResetGesture();
                }
            };
            touchCaptureHub.TouchCaptured += touchCapturedHandler;
            touchCaptureHub.RuntimeCaptureInterrupted += runtimeCaptureInterruptedHandler;
            Runtime.AppLifecycleStateChanged += appLifecycleStateChangedHandler;
            captureTask = Task.Run(() => RunCaptureLoopAsync(captureCts.Token));
        }
    }

    public async Task StopAsync(CancellationToken cancellationToken)
    {
        TouchCaptureHub? touchCaptureHub;
        EventHandler<TouchCapturedEventArgs>? touchCapturedHandler;
        EventHandler? runtimeCaptureInterruptedHandler;
        EventHandler<AppLifecycleStateChangedEventArgs>? appLifecycleStateChangedHandler;
        CancellationTokenSource? captureCts;
        Task? captureTask;

        lock (stateLock)
        {
            touchCaptureHub = this.touchCaptureHub;
            touchCapturedHandler = this.touchCapturedHandler;
            runtimeCaptureInterruptedHandler = this.runtimeCaptureInterruptedHandler;
            appLifecycleStateChangedHandler = this.appLifecycleStateChangedHandler;
            captureCts = this.captureCts;
            captureTask = this.captureTask;
            this.touchCaptureHub = null;
            this.touchCapturedHandler = null;
            this.runtimeCaptureInterruptedHandler = null;
            this.appLifecycleStateChangedHandler = null;
            this.captureCts = null;
            this.captureTask = null;
            gestureId = null;
            activePointerIds.Clear();
            pendingTrigger = null;
        }

        if (touchCaptureHub is not null && touchCapturedHandler is not null)
        {
            touchCaptureHub.TouchCaptured -= touchCapturedHandler;
        }
        if (touchCaptureHub is not null && runtimeCaptureInterruptedHandler is not null)
        {
            touchCaptureHub.RuntimeCaptureInterrupted -= runtimeCaptureInterruptedHandler;
        }
        if (appLifecycleStateChangedHandler is not null)
        {
            Runtime.AppLifecycleStateChanged -= appLifecycleStateChangedHandler;
        }

        captureCts?.Cancel();
        await WaitForTaskAsync(captureTask);
        captureCts?.Dispose();
    }

    public void Dispose()
    {
        if (disposed)
        {
            return;
        }

        disposed = true;
        if (touchCaptureHub is not null && touchCapturedHandler is not null)
        {
            touchCaptureHub.TouchCaptured -= touchCapturedHandler;
        }
        StopAsync(CancellationToken.None).GetAwaiter().GetResult();
        signal.Dispose();
    }

    private void Observe(CapturedTouch touch)
    {
        lock (stateLock)
        {
            if (captureCts is null || captureCts.IsCancellationRequested)
            {
                return;
            }

            switch (touch.Action)
            {
                case CapturedTouchAction.Down:
                    var beginsGesture = activePointerIds.Count == 0;
                    activePointerIds.Add(touch.PointerId);
                    if (beginsGesture)
                    {
                        gestureId = $"gesture-{Guid.NewGuid():N}";
                    }

                    EnqueueLocked(CreateTrigger(
                        touch,
                        beginsGesture
                            ? TouchVisualTreeGesturePhase.Started
                            : TouchVisualTreeGesturePhase.Checkpoint));
                    break;
                case CapturedTouchAction.Move:
                    activePointerIds.Add(touch.PointerId);
                    break;
                case CapturedTouchAction.Up:
                    activePointerIds.Remove(touch.PointerId);
                    EnqueueLocked(CreateTrigger(
                        touch,
                        activePointerIds.Count == 0
                            ? TouchVisualTreeGesturePhase.Ended
                            : TouchVisualTreeGesturePhase.Checkpoint));
                    break;
                case CapturedTouchAction.Cancel:
                    activePointerIds.Clear();
                    gestureId = null;
                    break;
            }
        }
    }

    private void ResetGesture()
    {
        lock (stateLock)
        {
            gestureId = null;
            activePointerIds.Clear();
            pendingTrigger = null;
        }
    }

    private async Task RunCaptureLoopAsync(CancellationToken cancellationToken)
    {
        var minimumIntervalTimestampTicks = Math.Max(
            1,
            (long)Math.Ceiling(minimumCaptureInterval.TotalSeconds * Stopwatch.Frequency));
        long nextCaptureAllowedTimestamp = 0;

        while (!cancellationToken.IsCancellationRequested)
        {
            try
            {
                await signal.WaitAsync(cancellationToken);
            }
            catch (OperationCanceledException)
            {
                return;
            }

            while (!cancellationToken.IsCancellationRequested)
            {
                bool hasPendingTrigger;
                lock (stateLock)
                {
                    hasPendingTrigger = pendingTrigger is not null;
                }

                if (!hasPendingTrigger)
                {
                    break;
                }

                var remainingTimestampTicks = nextCaptureAllowedTimestamp - Stopwatch.GetTimestamp();
                if (remainingTimestampTicks > 0)
                {
                    try
                    {
                        var delay = TimeSpan.FromSeconds(
                            remainingTimestampTicks / (double)Stopwatch.Frequency);
                        await Task.Delay(delay, cancellationToken);
                    }
                    catch (OperationCanceledException)
                    {
                        return;
                    }
                }

                TouchVisualTreeCaptureTrigger? trigger;
                lock (stateLock)
                {
                    trigger = pendingTrigger;
                    pendingTrigger = null;
                }

                if (trigger is null)
                {
                    continue;
                }

                try
                {
                    await captureAsync(trigger, cancellationToken);
                    nextCaptureAllowedTimestamp = Stopwatch.GetTimestamp() + minimumIntervalTimestampTicks;
                }
                catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
                {
                    return;
                }
                catch (Exception ex)
                {
                    Logger.Warning($"Touch visual-tree capture skipped: {ex.Message}");
                }
            }
        }
    }

    private TouchVisualTreeCaptureTrigger CreateTrigger(
        CapturedTouch touch,
        TouchVisualTreeGesturePhase phase)
    {
        return new TouchVisualTreeCaptureTrigger(
            gestureId ?? $"gesture-{Guid.NewGuid():N}",
            touch.Action,
            phase,
            touch.CapturedAtUtc);
    }

    private void EnqueueLocked(TouchVisualTreeCaptureTrigger trigger)
    {
        pendingTrigger = SelectPendingTrigger(pendingTrigger, trigger);
        if (signal.CurrentCount == 0)
        {
            try
            {
                signal.Release();
            }
            catch (SemaphoreFullException)
            {
            }
        }
    }

    private static TouchVisualTreeCaptureTrigger SelectPendingTrigger(
        TouchVisualTreeCaptureTrigger? pending,
        TouchVisualTreeCaptureTrigger incoming)
    {
        if (pending is null || incoming.GesturePhase == TouchVisualTreeGesturePhase.Started)
        {
            return incoming;
        }

        if (pending.GesturePhase == TouchVisualTreeGesturePhase.Started)
        {
            return pending;
        }

        return incoming.GesturePhase == TouchVisualTreeGesturePhase.Ended
            ? incoming
            : pending;
    }

    private static async Task WaitForTaskAsync(Task? task)
    {
        if (task is null)
        {
            return;
        }

        try
        {
            await task;
        }
        catch (OperationCanceledException)
        {
        }
        catch
        {
        }
    }
}
